CARDS-1933: PREMs Email notification Scheduling

acrowthe opened the pull request:
Adapt PROMs scheduling to PREMs use case
- Remove AUDIT-C special handling (now unneeded, can be replaced by reference questions)
- Add option to check frequency across different clinics for the same form
- Add conflict questionnaires to the questionnaire set

Test in PREMs run mode. Verify that patients will only receive a single OED or OAIP form every 26 weeks, regardless of which form they received previously and which clinic the visits are for.

acrowthe committed:
CARDS-1933: PREMs Email notification Scheduling
Adapt PROMs scheduling to PREMs use case
- Remove AUDIT-C special handling (now unneeded, can be replaced by reference questions)
- Add option to check frequency across different clinics for the same form
- Add conflict questionnaires to the questionnaire set

veronikaslc commented on modules/patient-portal/src/main/java/io/uhndata/cards/patients/internal/VisitChangeListener.java:
```diff
@@ -347,15 +359,19 @@ private Map<String, QuestionnaireFrequency> getQuestionnaireSetInformation(final
                 && "Visit".equals(this.subjectTypeUtils.getLabel(this.subjectUtils.getType(otherVisit)))) {
-                pruneQuestionnaireSetByVisit(otherVisit, visitInformation, questionnaireSetInfo);
+                pruneQuestionnaireSetByVisit(
+                    otherVisit, visitInformation, questionnaireSetInfo);
```
super-minor comment: this new line may be an unnedet change

IntegralProgrammer commented on modules/patient-portal/src/main/java/io/uhndata/cards/patients/internal/VisitChangeListener.java:
```diff
@@ -731,7 +698,7 @@ private static boolean isSpecificQuestionnaire(final Node questionnaire, final S
         public boolean hasRequiredInformation()
         {
-            return this.visitDate != null && StringUtils.isNotBlank(this.questionnaireSet);
+            return !!(this.visitDate != null && StringUtils.isNotBlank(this.questionnaireSet));
```
What is the purpose of this change?

acrowthe replied:
At some point during debugging I added this change to address what I though was a type casting issue. I'll verify if it is still needed or not

IntegralProgrammer commented on modules/patient-portal/src/main/java/io/uhndata/cards/patients/internal/VisitChangeListener.java:
```diff
@@ -308,31 +309,42 @@ private static Calendar addDays(final Calendar date, final int days)
      * @param session a service session providing access to the repository
      * @return a map of all questionnaires with frequency in the questionnaire set, keyed by questionnaire uuid
```
This doesn't return a `Map<String, QuestionnaireFrequency>` anymore but rather this returns a `QuestionnaireSetInfo` object.

IntegralProgrammer commented on modules/patient-portal/src/main/java/io/uhndata/cards/patients/internal/VisitChangeListener.java:
```diff
@@ -394,11 +410,13 @@ private void pruneQuestionnaireSetByVisit(final Node visit, final VisitInformati
+
         // Ignore forms for different clinics, or forms without a clinic
-        if (StringUtils.isBlank(visitClinic) || !visitInformation.getClinicPath().equals(visitClinic)) {
-            return;
-        } else {
+        if (questionnaireSetInfo.shouldIgnoreClinic()
+            || (!StringUtils.isBlank(visitClinic) && visitInformation.getClinicPath().equals(visitClinic))) {
             removeQuestionnairesFromVisit(visitInformation, visitDate, questionnaires, questionnaireSetInfo);
+        } else {
+            return;
         }
```
This is a bit difficult for me to follow. The comment says `Ignore forms for different clinics, or forms without a clinic`. So I see that in the if-statement, `removeQuestionnairesFromVisit` is called when `questionnaireSetInfo.shouldIgnoreClinic()` return `true`. If, however, `questionnaireSetInfo.shouldIgnoreClinic()` returns `false` and `visitClinic` is equal to `visitInformation.getClinicPath()` (which is non-empty), the statement is evaluated as follows:

`(false || (!false && true))`

which simplifies to

`(false || (true && true))`

which simplifies to `(true && true)` which is equal to `true` and thus `removeQuestionnairesFromVisit` is called. Yet the comment seems to indicate that Forms are ignored when `visitInformation.getClinicPath()` is _NOT_ equal to `visitClinic`.

acrowthe replied:
Code was correct, comment was misleading. Comment has been improved to hopefully explain the situation better.

IntegralProgrammer commented on modules/patient-portal/src/main/java/io/uhndata/cards/patients/internal/VisitChangeListener.java:
```diff
@@ -413,22 +431,30 @@ private void pruneQuestionnaireSetByVisit(final Node visit, final VisitInformati
+            Integer frequencyPeriod = null;
+            if (questionnaireSetInfo.getConflicts().containsKey(questionnaireIdentifier)) {
+                frequencyPeriod = questionnaireSetInfo.getConflicts().get(questionnaireIdentifier);
+            } else if (questionnaireSetInfo.frequency.containsKey(questionnaireIdentifier)) {
+                frequencyPeriod = questionnaireSetInfo.frequency.get(questionnaireIdentifier).getFrequency();
```
Why `questionnaireSetInfo.frequency` and not `questionnaireSetInfo.getFrequency()`?

IntegralProgrammer commented on modules/patient-portal/src/main/java/io/uhndata/cards/patients/internal/VisitChangeListener.java:
```diff
@@ -765,6 +732,40 @@ public String getClinicPath()
         }
     }
 
+    private static final class QuestionnaireSetInfo
+    {
+        private final Map<String, Integer> conflicts;
+        private final Map<String, QuestionnaireFrequency> frequency;
+        private boolean ignoreClinic;
+
+        QuestionnaireSetInfo()
+        {
+            this.conflicts = new HashMap<>();
+            this.ignoreClinic = false;
+            this.frequency = new HashMap<>();
+        }
+
+        public Map<String, Integer> getConflicts()
+        {
+            return this.conflicts;
+        }
+
+        public Map<String, QuestionnaireFrequency> getFrequency()
+        {
+            return this.frequency;
+        }
+
+        public boolean shouldIgnoreClinic()
+        {
+            return this.ignoreClinic;
+        }
+
+        public void setShouldIgnoreClinic(boolean ignoreClinic)
+        {
+            this.ignoreClinic = ignoreClinic;
+        }
+    }
```
By adding some more methods to this class and expanding some of the methods that are already there, we can simplify the interaction with the instances of the class. For example, we can have an `addConflict` method which adds a mapping to `this.conflicts` without needing to call `.getConflicts().put()`. Similarly, we can have a `hasConflict` method to replace the calling of `.getConflicts.containsKey()`. As well, the `getConflict` and `getFrequency` methods should be overloaded so that they can be passed a key and the corresponding value returned, or null if no such key exists.

acrowthe committed:
CARDS-1933: PREMs Email notification Scheduling
Improve code readability and naming

acrowthe committed:
CARDS-1933: PREMs Email notification Scheduling
Improve code clarity and comments

acrowthe committed:
CARDS-1933: PREMs Email notification Scheduling
Clean up unneeded boolean cast

IntegralProgrammer commented:
This might be out-of-scope for this PR, but shouldn't https://github.com/data-team-uhn/cards/blob/da600d49d626c92ff4c8b07e11ef621498efbc1e/modules/patient-portal/src/main/java/io/uhndata/cards/patients/internal/VisitChangeListener.java#L145 be placed inside a `finally` block to avoid leaking resources in case any previous statements in the `try` block throw exceptions?

marta- commented on prems-resources/clinical-data/src/main/resources/SLING-INF/content/Survey/AIP.xml:
```diff
@@ -25,6 +25,39 @@
+        <value>True</value>
+        <type>Boolean</type>
+    </property>
+    <node>
+        <name>conflict1</name>
+        <primaryNodeType>cards:QuestionnaireConflict</primaryNodeType>
+        <property>
+            <name>questionnaire</name>
+            <value>/Questionnaires/OAIP</value>
+            <type>Reference</type>
+        </property>
+        <property>
+            <name>frequency</name>
+            <value>26</value>
+            <type>Long</type>
+        </property>
+    </node>
+    <node>
+        <name>conflict2</name>
+        <primaryNodeType>cards:QuestionnaireConflict</primaryNodeType>
+        <property>
+            <name>questionnaire</name>
+            <value>/Questionnaires/OED</value>
+            <type>Reference</type>
+        </property>
+        <property>
+            <name>frequency</name>
+            <value>26</value>
+            <type>Long</type>
+        </property>
+    </node>
```
It would be nice to have an `any` option.

acrowthe committed:
CARDS-1933: PREMs Email notification Scheduling
Add conflit mode "any" to future proof prems clinics

acrowthe committed:
CARDS-1933: PREMs Email notification Scheduling
Configure recently created questionnaire sets for conflicts

acrowthe committed:
CARDS-1933: PREMs Email notification Scheduling
Fix configuration for new questionnaire sets

marta- commented:
Rebased on the latest dev

sashaandjic reviewed (approved):
tested

veronikaslc commented:
Comment not related to changes in this PR, rather observation:
I'm personally not a fan of overcomplicating things like [15 lines of extra code](https://github.com/data-team-uhn/cards/blob/155bb2e2bc5db0accf10c67fce9b0e7d3870e67c/modules/patient-portal/src/main/java/io/uhndata/cards/patients/internal/VisitChangeListener.java#L643-L658) just for one local operation of simple string comparison, given `isSpecificQuestionnaire()` not used more than once anywhere else even within the class.
If keeping, it would be more useful to inject it into the `QuestionnaireUtils` and reuse in places like [this](https://github.com/data-team-uhn/cards/blob/dev/proms-resources/permissions/src/main/java/io/uhndata/cards/proms/permissions/UnsubmittedFormsRestrictionPattern.java#L107-L110).

veronikaslc commented on modules/patient-portal/src/main/java/io/uhndata/cards/patients/internal/VisitChangeListener.java:
```diff
@@ -765,13 +734,120 @@ public String getClinicPath()
+                int frequencyPeriod;
+                if (CONFLICT_ANY.equals(this.conflictMode)) {
+                    frequencyPeriod = 0;
+                    for (final QuestionnaireRef ref : this.members.values()) {
```
```suggestion
                    frequencyPeriod = this.members.values().stream().map(ref -> ref.getFrequency()).max(Integer::compare).get();
```

veronikaslc replied:
something along those lines could look more compact

veronikaslc commented on modules/patient-portal/src/main/java/io/uhndata/cards/patients/internal/VisitChangeListener.java:
```diff
@@ -413,22 +440,29 @@ private void pruneQuestionnaireSetByVisit(final Node visit, final VisitInformati
      * @throws RepositoryException if iterating the patient's visits fails
```
It doesn't.

veronikaslc commented on modules/patient-portal/src/main/java/io/uhndata/cards/patients/internal/VisitChangeListener.java:
```diff
@@ -132,7 +135,7 @@ private void handleEvent(final ResourceChange event)
             final Node questionnaire = this.formUtils.getQuestionnaire(form);
```
```suggestion
            final Node questionnaire = this.formUtils.getQuestionnaire(form);
            final Node subject = this.formUtils.getSubject(form);
            if (isVisitInformation(questionnaire)) {
                // Create any forms that need to be created for this visit
                handleVisitInformationForm(form, subject, questionnaire, session);
            } else {
                final String subjectType = this.subjectTypeUtils.getLabel(this.subjectUtils.getType(subject));
                if ("Visit".equals(subjectType)) {
                    // Not a new visit information form, but it is a form for a visit.
                    // Check if all forms for the current visit are complete.
                    handleVisitDataForm(subject, session);
                }
            }
```

veronikaslc replied:
Just to save some coputational power

veronikaslc replied:
Also `subjectType` could be `null` so there is a risk of `NPE` at `"Visit".equals(subjectType)`

acrowthe replied:
The argument of "String".equals(String arg) is null safe so there is no risk of a `NPE` here as "Visit" is never null.

veronikaslc commented on modules/patient-portal/src/main/java/io/uhndata/cards/patients/internal/VisitChangeListener.java:
```diff
@@ -186,18 +189,18 @@ private void handleVisitInformationForm(final Node form, final Node visit, final
             return;
         }
 
         // Prune the quesionnaires to be created based on frequency.
```
```suggestion
        // Prune the questionnaires to be created based on frequency.
```

veronikaslc commented on modules/patient-portal/src/main/java/io/uhndata/cards/patients/internal/VisitChangeListener.java:
```diff
@@ -254,33 +259,33 @@ private void handleVisitDataForm(final Node visitSubject, final Session session)
                     }
```
`surveysCompleted != null` above looks redundant

acrowthe replied:
The null check is required or the `==` can throw a null pointer exception

marta- replied:
No, `==` never throws NPE as it is an operator that compares references, not a method that can be called for an object.

marta- replied:
But here we have another instance of using `==` to compare an object to a constant, which may or may not produce the expected result. The null check should be kept, and the second `==` should be replaced with `.equals`.

acrowthe replied:
I thought so too but removing the null check leads to a null pointer exception on that line. It must be due to the comparison between nullable `Long` and primitive non-nullable `long`

acrowthe replied:
Fixed to use .equals()

sdumitriu replied:
The code is good as it is. `surveysCompleted` is a java object of type `Long`, not a primitive `long`. `surveysCompleted == 1L` forces it into a primitive value, which behind the scenes invokes a method on the object (`surveysCompleted.longValue()`), which could cause a NPE if the object is null. Once the primitive `long` value is obtained out of it, the comparison with `==` is safe, since it compares two primitive values.

An alternative way of writing this safely is `Long.valueOf(1L).equals(surveysCompleted)`, which also takes care of the null case, but it's not really an improvement over the current code.

marta- replied:
via @sdumitriu : In fact this line is perfect the way it is. `== 1L` triggers a method call to pull the primitive value to be compared to `1L` (also a primitive value). Because there's an implicit method call, the null check is necessary. And obviously comparing two primitive values with `==` is safe.

marta- replied:
@acrowthe please revert the latest commit and sorry for the noise.

acrowthe replied:
Done

veronikaslc reviewed (commented):
Most of codereview comments not related to changes in this PR but rather to the general `VisitChangeListener` code.

acrowthe committed:
CARDS-1933: PREMs Email notification Scheduling
- Clean up to address code review comments

acrowthe committed:
CARDS-1933: PREMs Email notification Scheduling
Fix comment typo

veronikaslc commented on modules/patient-portal/src/main/java/io/uhndata/cards/patients/internal/VisitChangeListener.java:
```diff
@@ -765,13 +723,118 @@ public String getClinicPath()
+                int frequencyPeriod;
+                if (CONFLICT_ANY.equals(this.conflictMode)) {
+                    frequencyPeriod = 0;
+                    frequencyPeriod = this.members.values().stream()
```
Those 2 lines look strange together (757 and 758). 
May be initiate `frequencyPeriod` to `0` at line 755?

acrowthe committed:
CARDS-1933: PREMs Email notification Scheduling
Clean up frequency configuration and assignment

sashaandjic reviewed (approved):
tested

sdumitriu commented on modules/patient-portal/src/main/resources/SLING-INF/nodetypes/patient-portal.cnd:
```diff
@@ -87,6 +110,15 @@
+  // If true, apply the questionnaire's frequency across all clinics for these questionnaires
+  - frequencyIgnoreClinic (boolean)
+
+  // How any questionnaires not included in this set should be checked for conflicts
```
Please document what values are accepted for this property.

veronikaslc commented on modules/patient-portal/src/main/java/io/uhndata/cards/patients/internal/VisitChangeListener.java:
```diff
@@ -299,136 +306,164 @@ private static Calendar addDays(final Calendar date, final int days)
+                } else if (node.isNodeType("cards:QuestionnaireConflict")) {
+                    if (node.hasProperty("questionnaire")) {
+                        final String uuid = node.getProperty("questionnaire").getString();
+                        final int frequency = (int) node.getProperty("frequency").getLong();
```
Just noticed that in one case we initiate `frequency` to 0 on line 329 and then check `if (node.hasProperty("frequency"))` on line 330, but here we don't. Shell we?

sdumitriu replied:
No, for `cards:QuestionnaireConflict` the frequency is mandatory, while for `cards:QuestionnaireRef` it is optional.

sdumitriu commented on modules/patient-portal/src/main/resources/SLING-INF/nodetypes/patient-portal.cnd:
```diff
@@ -26,6 +26,29 @@
+  - sling:resourceType (STRING) = "cards/QuestionnaireConflict" mandatory autocreated protected
+
+  // Hardcode the resource supertype.
+  - sling:resourceSuperType (STRING) = "cards/Resource" mandatory autocreated protected
```
I don't think this is a resource. A `cards/Resource` is something that can be displayed by itself in the UI, that is stored under a `ResourceHomepage`, and that can be listed in a livetable.

```suggestion
```

acrowthe committed:
CARDS-1933: PREMs Email notification Scheduling
- Improve documentation on conflict modes
- Remove incorrect "Resource" supertype from conflict node type
